lib/matplotlib/backend_bases.py

```diff
@@ -3195,3 +3207,560 @@ def zoom(self, *args):
+
+    def set_tool_keymap(self, name, *keys):
+        """
+        Set the keymap to associate with the specified tool
+
+        Parameters
+        ----------
+        name : string
+            Name of the Tool
+        keys : keys to associate with the Tool
+        """
+
+        if name not in self._tools:
+            raise KeyError('%s not in Tools' % name)
+
+        self._remove_keys(name)
+
+        for key in keys:
+            for k in validate_stringlist(key):
+                if k in self._keys:
+                    warnings.warn('Key %s changed from %s to %s' %
+                                  (k, self._keys[k], name))
+                self._keys[k] = name
+
+    def remove_tool(self, name):
+        """
+        Remove tool from `Navigation`
+
+        Parameters
+        ----------
+        name : string
+            Name of the Tool
+        """
+
+        tool = self.get_tool(name)
+        tool.destroy()
+
+        # If is a toggle tool and toggled, untoggle
+        if getattr(tool, 'toggled', False):
+            self.tool_trigger_event(tool, 'navigation')
+
+        self._remove_keys(name)
+
+        s = 'tool_removed_event'
+        event = ToolEvent(s, self, tool)
+        self._callbacks.process(s, event)
+
+        del self._tools[name]
+
+    def add_tools(self, tools):
+        """
+        Add multiple tools to `NavigationBase`
+
+        Parameters
+        ----------
+        tools : {str: class_like}
+            The tools to add in a {name: tool} dict, see `add_tool` for more
+            info.
+        """
+
+        for name, tool in six.iteritems(tools):
+            self.add_tool(name, tool)
+
+    def add_tool(self, name, tool, *args, **kwargs):
+        """
+        Add tool to `NavigationBase`
+
+        Add a tool to the tools controlled by Navigation
+
+        If successful adds a new event `tool_trigger_name` where **name** is
+        the **name** of the tool, this event is fired everytime
+        the tool is triggered.
+
+        Parameters
+        ----------
+        name : str
+            Name of the tool, treated as the ID, has to be unique
+        tool : class_like, i.e. str or type
+            Reference to find the class of the Tool to added.
+
+        Notes
+        -----
+        args and kwargs get passed directly to the tools constructor.
+
+        See Also
+        --------
+        matplotlib.backend_tools.ToolBase : The base class for tools.
+        """
+
+        tool_cls = self._get_cls_to_instantiate(tool)
+        if tool_cls is False:
+            raise ValueError('Impossible to find class for %s' % str(tool))
+
+        if name in self._tools:
+            warnings.warn('A "Tool class" with the same name already exists, '
+                          'not added')
+            return self._tools[name]
+
+        self._tools[name] = tool_cls(self, name, *args, **kwargs)
+
+        if tool_cls.keymap is not None:
+            self.set_tool_keymap(name, tool_cls.keymap)
```

Comment:
@tacaswell ``set_tool_keymap(name, keymap)`` does per-instance key binding, it sets the keymap of the instance identified by ``name`` to the specified keymap.  Here it takes the default keymap (a class level property) and binds it to the instance.  Does that make it clearer for you?